File: src/main/java/es/xan/servantv3/outlet/OutletVerticle.java
package es.xan.servantv3.outlet;

import es.xan.servantv3.AbstractServantVerticle;
import es.xan.servantv3.Action;
import es.xan.servantv3.Constant;
import es.xan.servantv3.MessageBuilder;
import es.xan.servantv3.MessageBuilder.ReplyBuilder;
import es.xan.servantv3.SSHUtils;
import es.xan.servantv3.messages.Configure;
import es.xan.servantv3.messages.UpdateState;
import io.vertx.core.eventbus.Message;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;

public class OutletVerticle extends AbstractServantVerticle {

	private static final Logger LOGGER = LoggerFactory.getLogger(OutletVerticle.class);
	
	public OutletVerticle() {
		super(Constant.OUTLET_VERTICLE);
		
		supportedActions(Actions.values());
	}

	private String mHost;
	private String mLogin;
	private String mPassword;
	
	private static String ON_COMMAND = "echo \"1\" > /proc/power/relay1";
	private static String OFF_COMMAND = "echo \"0\" > /proc/power/relay1";
	
	public enum Actions  implements Action {
		SWITCHER(UpdateState.class),
		SET(Configure.class)
		;
		
		Class<?> beanClass;
		
		private Actions(Class<?> beanClass) {
			this.beanClass = beanClass;
		}
		
		@Override
		public Class<?> getPayloadClass() {
			return beanClass;
		}
		
	}
	
	public void start() {
		super.start();
		loadConfiguration(vertx.getOrCreateContext().config().getJsonObject("OutletVerticle"));
	}
	
	private void loadConfiguration(JsonObject config) {
		mHost = config.getString("host", "192.168.1.100");
		mLogin = config.getString("login", "ubnt");
		mPassword = config.getString("password", "ubnt");
	}
	
	public void set(Configure conf, Message<Object> message) {
		ReplyBuilder builderOn = MessageBuilder.createReply();
		switch (conf.getField()) {
		case "host":
			this.mHost = conf.getValue();
			builderOn.setOk();
			builderOn.setMessage(conf.getValue());
			break;
		case "login":
			this.mLogin= conf.getValue();
			builderOn.setOk();
			builderOn.setMessage(conf.getValue());
			break;
		case "password":
			this.mPassword= conf.getValue();
			builderOn.setOk();
			builderOn.setMessage(conf.getValue());
			break;
		default:
			builderOn.setError();
			builderOn.setMessage("field:" + conf.getField() + " value" + conf.getValue());
			break;
		}
		message.reply(builderOn.build());
	}
	
	public void switcher(UpdateState switcher, Message<Object> message) {
		boolean ok = false;
		try {
			if ("on".equals(switcher.getNewStatus())) {
				ok = SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, ON_COMMAND);
			} else {
				ok = SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, OFF_COMMAND);
			}
		} catch (Exception e) {
			LOGGER.warn(e.getMessage(), e);
		} finally {
			ReplyBuilder builderOn = MessageBuilder.createReply();
			if (ok) {
				builderOn.setOk();
			} else {
				builderOn.setError();
			}
			message.reply(builderOn.build());
		}
	}
}

File: src/main/kotlin/es/xan/servantv3/outlet/OutletVerticle.kt
package es.xan.servantv3.outlet

import es.xan.servantv3.AbstractServantVerticle
import es.xan.servantv3.Constant
import es.xan.servantv3.Action
import es.xan.servantv3.messages.UpdateState
import es.xan.servantv3.messages.Configure
import io.vertx.core.eventbus.Message
import es.xan.servantv3.SSHUtils
import es.xan.servantv3.MessageBuilder
import io.vertx.core.logging.LoggerFactory

class OutletVerticle : AbstractServantVerticle(Constant.OUTLET_VERTICLE) {
	companion object {
        val LOG = LoggerFactory.getLogger(OutletVerticle::class.java.name) 
    }
	
	init {
		supportedActions(Actions::class.java)
	}
	
	enum class Actions(val clazz : Class<*>? ) : Action {
		SWITCHER(UpdateState::class.java),
		STATUS(null),
		SET(Configure::class.java);
	}
	
	val mHost		: String by lazy { config().getJsonObject("OutletVerticle").getString("host", "192.168.1.100") }
	val mLogin		: String by lazy { config().getJsonObject("OutletVerticle").getString("login", "ubnt") }
	val mPassword	: String by lazy { config().getJsonObject("OutletVerticle").getString("password", "ubnt") }
	
	val ON_COMMAND	= "echo \"1\" > /proc/power/relay1"
	val OFF_COMMAND	= "echo \"0\" > /proc/power/relay1"
	val STATUS_COMMAND = "grep . /proc/power/*";
	
	fun switcher(switcher : UpdateState, message: Message<Any>) {
		val reply = MessageBuilder.createReply().apply {
			try {
				when (switcher.newStatus) {
					"on" -> SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, ON_COMMAND);
					else -> SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, OFF_COMMAND);
				}
				setOk();
			} catch (e: Throwable) {
				LOG.warn("Problems trying to manage ssh remote command", e);
				setError();
				setMessage(e.localizedMessage)
			}
		}
		
		message.reply(reply.build());
	}
	
	fun status(message: Message<Any>) {
		val reply = MessageBuilder.createReply().apply {
			try {
				val response = SSHUtils.runRemoteCommandExtended(mHost, mLogin, mPassword, STATUS_COMMAND);
				setOk()
				setMessage(response.output)
			} catch (e: Throwable) {
				setError()
				setMessage(e.message)
			}
		}
		
		message.reply(reply.build())
	}

}
